### PR TITLE
[peft] Port DoRA support

### PR DESCRIPTION
**Pull request by ananthsub**
Port code here:
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/peft/dora.py

Add unit + integration tests

```
pytest -svx tests/unit_tests/peft/test_dora_layers.py
pytest -svx tests/unit_tests/peft/test_dora.py
```
+ Update LoRA + DoRA transforms to be idempotent to cover edge case when transforms are applied over already transformed models

### CODE REVIEW

File: src/megatron/hub/peft/dora.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        dim (int): Dimension of the low-rank projection space. Defaults to 32.
+        alpha (int): Weighting factor for the low-rank projection. Defaults to 64.
+        dropout (float): Dropout rate for the low-rank projection. Defaults to 0.0.
+        dropout_position (Literal['pre', 'post'], optional): Position for applying dropout.
+            Can be 'pre' (before the low-rank projection) or 'post' (after). Defaults to 'pre'.
+        lora_A_init_method (str): Initialization method for the low-rank matrix A. Defaults to "xavier".
+        lora_B_init_method (str): Initialization method for the low-rank matrix B. Defaults to "zero".
+
+    Example:
+    --------
```
Comment (maanug-nv) [inline src/megatron/hub/peft/dora.py]: same docstring format concern as Lora. #84 should be merged soon. can run build locally for now to check for any other errors.
